### PR TITLE
Enable lang testcases depend on table type & related bug fixes

### PR DESCRIPTION
**Pull request by mohanvive**
## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 1]
```diff
@@ -1134,6 +1134,14 @@ public void setForeachTypedBindingPatternType(BLangForeach foreachNode) {
             case TypeTags.XML:
                 varType = BUnionType.create(null, symTable.xmlType, symTable.stringType);
                 break;
+            case TypeTags.TABLE:
+                BTableType tableType = (BTableType) collectionType;
+                if (tableType.constraint.tag == TypeTags.NONE) {
```
Comment (gimantha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:1139]: is it possible to have noType for the constraint?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 2]
```diff
@@ -1226,6 +1234,14 @@ public void setFromClauseTypedBindingPatternType(BLangFromClause fromClause) {
...
```
Comment (gimantha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:1239]: is it possible to have noType for the constraint?

File: langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/statements/foreach/ForeachTableTypedBindingPatternsTests.java [hunk 3]
```diff
@@ -31,43 +31,42 @@
...
         Assert.assertEquals(returns[0].stringValue(), "0:1:Mary:300.5 1:2:John:200.5 2:3:Jim:330.5 ");
     }
 
-    @Test (enabled = false)
-    public void testEmptyTableIteration() {
-        BValue[] returns = BRunUtil.invoke(program, "testEmptyTableIteration");
-        Assert.assertEquals(returns.length, 1);
-        Assert.assertEquals(returns[0].stringValue(), "");
-    }
+//    @Test
```
Comment (gimantha) [inline langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/statements/foreach/ForeachTableTypedBindingPatternsTests.java:66]: I will enable this.. I have the fix

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/types/BTableType.java [hunk 4]
```diff
@@ -56,12 +57,12 @@ public BType getKeyType() {
 
     @Override
     public <V> V getZeroValue() {
-        return null;
+        return (V) new TableValueImpl<String, V>(new BTableType(constraint));
```
Comment (gimantha) [inline bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/types/BTableType.java:60]: we cant have string here. key type can be any anydata type for tables.

File: langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/statements/foreach/ForeachTableTypedBindingPatternsTests.java (hunk 3 above)
Comment (mohanvive) [reply langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/statements/foreach/ForeachTableTypedBindingPatternsTests.java:66]: OK.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java (hunk 1 above)
Comment (mohanvive) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java]: Fixed.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java (hunk 2 above)
Comment (mohanvive) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java]: Fixed.

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/types/BTableType.java (hunk 4 above)
Comment (mohanvive) [reply bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/types/BTableType.java]: Fixed.

**Pull request merged**